File: WdtcCore/src/main/java/org/wdt/wdtc/manger/GameFileManger.java
package org.wdt.wdtc.manger;


import lombok.Getter;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.download.game.DownloadVersionGameFile;
import org.wdt.wdtc.game.GameVersionJsonObject;
import org.wdt.wdtc.utils.gson.JSONUtils;

import java.io.File;
import java.io.IOException;
import java.util.Calendar;
import java.util.Date;

@Getter
public class GameFileManger extends GameFolderManger {
    protected final String VersionNumber;

    public GameFileManger(String VersionNumber) {
        this.VersionNumber = VersionNumber;
    }

    public GameFileManger(String VersionNumber, File here) {
        super(here);
        this.VersionNumber = VersionNumber;
    }

    public static void downloadVersionManifestJsonFileTask() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(new Date());
        calendar.add(Calendar.DATE, -7);
        try {
            if (FileUtils.isFileNotExists(FileManger.getVersionManifestFile()) ||
                    FileUtils.isFileOlder(FileManger.getVersionManifestFile(), calendar.getTime())) {
                DownloadVersionGameFile.DownloadVersionManifestJsonFile();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    public File getVersionPath() {
        return new File(getGameVersionsPath(), VersionNumber);
    }

    public File getVersionJson() {
        return new File(getVersionPath(), VersionNumber + ".json");
    }

    public File getVersionJar() {
        return new File(getVersionPath(), VersionNumber + ".jar");
    }

    public File getVersionLog4j2() {
        return new File(getVersionPath(), "log4j2.xml");
    }

    public File getVersionNativesPath() {
        return new File(getVersionPath(), "natives-windows-x86_64");
    }

    public File getGameAssetsListJson() throws IOException {
        return new File(getGameAssetsdir(), "indexes/" + getGameVersionJsonObject().getAssets() + ".json");
    }

    public File getGameOptionsFile() {
        return new File(getVersionPath(), "options.txt");
    }

    public File getGameModsPath() {
        return new File(getVersionPath(), "mods");
    }

    public File getGameLogDir() {
        return new File(getVersionPath(), "logs");
    }

    public void PutToVersionJson(GameVersionJsonObject o) {
        JSONUtils.ObjectToJsonFile(getVersionJson(), o);
    }

    public GameVersionJsonObject getGameVersionJsonObject() throws IOException {
        return JSONUtils.JsonFileToClass(getVersionJson(), GameVersionJsonObject.class);
    }

    public File getVersionConfigFile() {
        return new File(getVersionPath(), "config.json");
    }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/manger/GameFileManger.kt
package org.wdt.wdtc.core.manger

import org.wdt.utils.gson.readFileToClass
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.utils.io.isFileNotExists
import org.wdt.utils.io.isFileOlder
import org.wdt.wdtc.core.download.game.DownloadVersionGameFile
import org.wdt.wdtc.core.game.GameVersionJsonObject
import org.wdt.wdtc.core.game.readGameVersionJsonObjectGson
import java.io.File
import java.io.IOException
import java.util.*


open class GameFileManger(
  val versionNumber: String,
  here: File = currentSetting.defaultGamePath
) : GameDirectoryManger(here) {
  
  val versionDirectory: File
    get() = File(gameVersionsDirectory, versionNumber)
  val versionJson: File
    get() = File(versionDirectory, "$versionNumber.json")
  val versionJar: File
    get() = File(versionDirectory, "$versionNumber.jar")
  val versionLog4j2: File
    get() = File(versionDirectory, "log4j2.xml")
  val versionNativesPath: File
    get() = File(versionDirectory, "natives-windows-x86_64")

  @get:Throws(IOException::class)
  val gameAssetsListJson: File
    get() = File(gameAssetsDirectory, "indexes/${gameVersionJsonObject.assets}.json")
  val gameOptionsFile: File
    get() = File(versionDirectory, "options.txt")
  val gameModsPath: File
    get() = File(versionDirectory, "mods")
  val gameLogDirectory: File
    get() = File(versionDirectory, "logs")

  fun GameVersionJsonObject.putToVersionJson() {
    versionJson.writeObjectToFile(this)
  }

  val gameVersionJsonObject: GameVersionJsonObject
    get() = versionJson.run {
      readFileToClass(readGameVersionJsonObjectGson)
    }

  val laucnherProfiles: File
    get() = File(gameDirectory, "Launcher_profiles.json")
  val versionConfigFile: File
    get() = File(versionDirectory, "config.json")
}


fun downloadVersionManifestJsonFileTask() {
  val calendar = Calendar.getInstance().apply {
    time = Date()
    add(Calendar.DATE, -7)
  }
  versionManifestFile.run {
    if (isFileNotExists() || isFileOlder(calendar.time)) {
      DownloadVersionGameFile.startDownloadVersionManifestJsonFile()
    }
  }
}

